tests/lib/rules/prefer-promise-reject-errors.js

```diff
@@ -46,7 +48,13 @@ ruleTester.run("prefer-promise-reject-errors", rule, {
 
         // Optional chaining
         "Promise.reject(obj?.foo)",
-        "Promise.reject(obj?.foo())"
+        "Promise.reject(obj?.foo())",
+
+        // Assignments
+        "Promise.reject(foo = new Error())",
+        "Promise.reject(foo ||= 5)",
+        "Promise.reject(foo.bar ??= 5)",
+        "Promise.reject(foo[bar] &&= 5)"
```

Comment:
Can you help me understand why these are considered valid? For `||=` and `??=`, if the left-hand side is an `Error`, it'll evaluate to an `Error`, so some but not all cases are okay. Perhaps a future improvement would flag the right-hand side? But for `&&=`, if I'm tracing correctly, it will never evaluate to an `Error` if the right-hand side is not an `Error`. For `&&=`, could we ignore the left side entirely and only look at the right side?

Reply:
> Can you help me understand why these are considered valid? For `||=` and `??=`, if the left-hand side is an `Error`, it'll evaluate to an `Error`, so some but not all cases are okay. Perhaps a future improvement would flag the right-hand side?

It's a bit weird, but `prefer-promise-reject-errors` and `no-throw-literal` allow all expressions that at least sometimes can evaluate to an `Error` object. We already noticed this in the optional chaining PR https://github.com/eslint/eslint/pull/13416#issuecomment-650828276

I think we should change this logic at some point (maybe in major version).

> But for &&=, if I'm tracing correctly, it will never evaluate to an Error if the right-hand side is not an Error. For &&=, could we ignore the left side entirely and only look at the right side?

Good catch! This doesn't seem right from any point of view.

Reply:
I'll fix `&&=` asap.

Reply:
@kaicataldo and I were chatting on Discord, and we can delay the release until Monday to give this more time to settle.

If `couldBeError()` is updated to treat `&&=` like `=`, `&&=` would be inconsistent with `&&`.

1. Could we update `&&` to only look at the right side in a semver-minor bug fix, or does it need to be a major?
1. If semver-minor, should it be in this PR or a separate one?
1. If semver-major, for the time being should `&&=` behave more like `=` only looking at the right side or `&&` looking at both? I'd argue for the correct behavior and accepting the inconsistency with `&&`.

Reply:
It's now updated to treat `&&=` like `=`, which is logically correct.

After this update, there will be an inconsistency between `&&=` and `&&`.

I think we could fix `&&` in a minor version.

It already looks like this rule should be reporting more than it's reporting now:  

> This rule aims to ensure that Promises are only rejected with Error objects.

The same goes for `no-throw-literal`:

> This rule is aimed at maintaining consistency when throwing exception by disallowing to throw literals and other expressions which cannot possibly be an Error object.

We could also consider updating these rules to report `foo || 5`, `foo ||= 5`, `foo ?? 5`, `foo ??= 5` and similar expressions in a major version.

Reply:
> If `couldBeError()` is updated to treat `&&=` like `=`, `&&=` would be inconsistent with `&&`.
> 
> 1. Could we update `&&` to only look at the right side in a semver-minor bug fix, or does it need to be a major?
> 2. If semver-minor, should it be in this PR or a separate one?
> 3. If semver-major, for the time being should `&&=` behave more like `=` only looking at the right side or `&&` looking at both? I'd argue for the correct behavior and accepting the inconsistency with `&&`.

I'd vote to update `&&` in a semver-minor bug fix, in a separate PR.

Reply:
I opened https://github.com/eslint/eslint/issues/13634 for the `&&` case, and I'm marking this as resolved because the `&&=` case is fixed.